parl/framework/algorithm.py

```diff
@@ -0,0 +1,187 @@
+
+    def _check_spec_duplicate_names(self):
+        """
+        Check if there are two specs that have different shapes or types but the same name.
+        Assume each element in specs is already unique, so we only have to find duplicate names.
+        """
+        specs = self.model.get_input_specs() \
+                + self.model.get_action_specs() \
+                + self.model.get_state_specs()
+        names = [name for name, _ in specs]
+        duplicates = set([n for n in names if names.count(n) > 1])
+        assert not duplicates, \
+            "duplicate names with different specs: " + " ".join(duplicates)
+
+    def _setup_programs(self):
+        self.predict_program = fluid.Program()
+        self.learn_program = fluid.Program()
+
+        ######## define the predict program  ########
+        fluid.switch_main_program(self.predict_program)
+        data_layer_dict = self._create_data_layers(self.model.get_input_specs(
+        ))
+        data_layer_dict.update(
+            self._create_data_layers(self.model.get_state_specs()))
+        self.predict_feed_names = sorted(data_layer_dict.keys())
+
+        inputs = [
+            data_layer_dict[name] for name, _ in self.model.get_input_specs()
+        ]
+        states = [
+            data_layer_dict[name] for name, _ in self.model.get_state_specs()
+        ]
+        result = self.model.predict(inputs, states)
+        assert len(result) == len(self.model.get_action_specs()) \
+            + len(self.model.get_state_specs())
+        actions = [(name, result[name])
+                   for name, _ in self.model.get_action_specs()]
+        states = [(name, result[name])
+                  for name, _ in self.model.get_state_specs()]
+        self.predict_fetch = [actions, states]
+
+        ######### define the learn program ###########
+        fluid.switch_main_program(self.learn_program)
+        data_layer_dict = self._create_data_layers(self.model.get_input_specs(
+        ))
+        data_layer_dict.update(
+            self._create_data_layers(self.model.get_state_specs()))
+        data_layer_dict.update(
+            self._create_data_layers(self.model.get_action_specs()))
+        self.learn_feed_names = sorted(data_layer_dict.keys())
+
+        inputs = [
+            data_layer_dict[name] for name, _ in self.model.get_input_specs()
+        ]
+        states = [
+            data_layer_dict[name] for name, _ in self.model.get_state_specs()
+        ]
+        actions = [
+            data_layer_dict[name] for name, _ in self.model.get_action_specs()
+        ]
+        ### TODO: implement a recurrent layer to strip the sequence information
+        result = self.model.learn(inputs, states, actions, self.referent_model)
+        assert len(result) == 1
+        self.avg_cost = layers.mean(x=result["cost"])
+        optimizer = fluid.optimizer.SGD(
+            learning_rate=self.hyperparas["learning_rate"])
+        optimizer.minimize(self.avg_cost)
```

Comment:
line 134-139 should be the learn() of the Algorithm I am thinking of. When user implement an new algorithm, he only needs to write the actual logic of the algorithm , should not care about things like line 88-132